pkg/controller/elasticsearch/driver/desired_nodes.go

```diff
@@ -0,0 +1,79 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License 2.0;
+// you may not use this file except in compliance with the Elastic License 2.0.
+
+package driver
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"go.elastic.co/apm"
+	corev1 "k8s.io/api/core/v1"
+
+	esv1 "github.com/elastic/cloud-on-k8s/pkg/apis/elasticsearch/v1"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/reconciler"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/tracing"
+	esclient "github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/client"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/nodespec"
+	"github.com/elastic/cloud-on-k8s/pkg/utils/k8s"
+)
+
+func (d *defaultDriver) updateDesiredNodes(
+	ctx context.Context,
+	k8sClient k8s.Client,
+	esClient esclient.Client,
+	esReachable bool,
+	expectedResources nodespec.ResourcesList,
+) *reconciler.Results {
+	span, ctx := apm.StartSpan(ctx, "update_desired_nodes", tracing.SpanTypeApp)
+	defer span.End()
+	results := &reconciler.Results{}
+	// We compute the desired nodes state to update the condition
+	var resourceNotAvailableErr *nodespec.ResourceNotAvailable
+	nodes, requeue, err := expectedResources.ToDesiredNodes(ctx, k8sClient, d.ES.Spec.Version)
+	switch {
+	case err == nil:
+		d.ReconcileState.ReportCondition(
+			esv1.ResourcesAwareManagement,
+			corev1.ConditionTrue,
+			fmt.Sprintf("Successfully calculated compute and storage resources from Elasticsearch resource generation %d", d.ES.Generation),
+		)
+	case errors.As(err, &resourceNotAvailableErr):
+		// It is not possible to build the desired node spec because of the Elasticsearch specification
+		d.ReconcileState.ReportCondition(
+			esv1.ResourcesAwareManagement,
+			corev1.ConditionFalse,
+			fmt.Sprintf("Cannot get compute and storage resources from Elasticsearch resource generation %d: %s", d.ES.Generation, err.Error()),
+		)
+		// It is fine to continue, error is only reported through the condition.
+		// We should however clear the desired nodes API since we are a degraded (not resources aware) mode.
+		if esReachable {
+			return results.WithError(esClient.DeleteDesiredNodes(ctx))
```

Comment:
> Is there an optimisation to say we only delete if there are desired nodes to begin with? 

This is a very good question as I think it actually raises another one: should we always call `DELETE /_internal/desired_nodes` whatever is the current history ID?

@fcofdez May I have your thoughts on this suggestion? My understanding is that the only way to get the current desired nodes state is by calling `/_internal/desired_nodes/_latest`. However, what if the `_latest` stored state is related to a previous history/instance of the cluster? Said differently, what if the cluster has been restored and is now running with a new history? Calling `DELETE /_internal/desired_nodes` without being able to specify an history ID would also delete that previous state, which is maybe not what we want? (maybe it is, just want to be sure it makes sense from an Elasticsearch pov to delete all the histories at once)

Reply:
> Is there an optimisation to say we only delete if there are desired nodes to begin with?

I think it's better to let Elasticsearch to handle that, it is almost a no-op when there are no desired nodes, so as long as we call that API occasionally it should be fine.

> This is a very good question as I think it actually raises another one: should we always call DELETE /_internal/desired_nodes whatever is the current history ID?

No, `DELETE /_internal/desired_nodes` is expected to be used when the orchestrator decides to move to the "unmanaged" mode, meaning that Elasticsearch would stop relying on the desired nodes information to make decisions.

> My understanding is that the only way to get the current desired nodes state is by calling /_internal/desired_nodes/_latest. 

You're correct, the only way to get the current desired nodes is by calling `GET /_internal/desired_nodes/_latest`.

> However, what if the _latest stored state is related to a previous history/instance of the cluster? Said differently, what if the cluster has been restored and is now running with a new history? Calling DELETE /_internal/desired_nodes without being able to specify an history ID would also delete that previous state, which is maybe not what we want? (maybe it is, just want to be sure it makes sense from an Elasticsearch pov to delete all the histories at once)

For the restore scenario that you're describing, if the cluster state is restored, the latest desired nodes might be a few versions behind from the latest desired nodes version that the operator is aware of. As long as the next update uses a version greater than the restored desired nodes, the update should proceed without any issues. In any case, Elasticsearch returns an error if the provided version is behind the latest stored version.

Reply:
> No, DELETE /_internal/desired_nodes is expected to be used when the orchestrator decides to move to the "unmanaged" mode, meaning that Elasticsearch would stop relying on the desired nodes information to make decisions.

Agreed. In ECK this will happen if the resources cannot be calculated (for example if no CPU has been provided). This however might be an involuntary situation, for example if the user makes a mistake, in that case all the "histories" will be deleted. We can revisit this mechanism once the API is more lenient regarding missing resources.

> I think it's better to let Elasticsearch to handle that, it is almost a no-op when there are no desired nodes, so as long as we call that API occasionally it should be fine.

👍 , thanks !